feat(ast): add `ThisExpression` to `TSTypeName`

Boshen opened the pull request:
part of #12042

Aligned with estree: https://github.com/typescript-eslint/typescript-eslint/blob/5c5e3d5c3853ab613e06be0d77a40e970017b3fc/packages/ast-spec/src/unions/EntityName.ts#L5

Boshen commented:
* **#12156** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12156?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12156?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12156).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented on tasks/coverage/snapshots/semantic_babel.snap:
```diff
@@ -191,16 +191,6 @@ Unresolved references mismatch:
-semantic Error: tasks/coverage/babel/packages/babel-parser/test/fixtures/estree/typescript/import-dot-this/input.js
-Unresolved references mismatch:
-after transform: ["this"]
-rebuilt        : []
```
:-)

Dunqing commented on crates/oxc_transformer/src/decorator/legacy/metadata.rs:
```diff
@@ -419,6 +419,7 @@ impl<'a> LegacyDecoratorMetadata<'a, '_> {
                     Some(ctx.ast.expression_logical(SPAN, left, LogicalOperator::And, member))
                 }
             }
+            TSTypeName::ThisExpression(_) => None,
```
Ok, I will take a look later.

Dunqing replied:
It is correct. It would enter 
https://github.com/oxc-project/oxc/blob/35090dc8720a9924b422f7be1efa0b706b177d74/crates/oxc_transformer/src/decorator/legacy/metadata.rs#L374-L376

this branch before, so the behavior of returning `None` is the same as before. I added a [test](https://github.com/oxc-project/oxc/pull/12156/commits/35090dc8720a9924b422f7be1efa0b706b177d74) to cover this.

Boshen commented:
### Merge activity

* **Jul 13, 1:35 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Boshen committed:
feat(ast): add `ThisExpression` to `TSTypeName` (#12156)

part of #12042

Aligned with estree: https://github.com/typescript-eslint/typescript-eslint/blob/5c5e3d5c3853ab613e06be0d77a40e970017b3fc/packages/ast-spec/src/unions/EntityName.ts#L5
